x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/apikey/ApiKey.java

```diff
@@ -53,7 +57,34 @@ public ApiKey(
         String username,
         String realm,
         @Nullable Map<String, Object> metadata,
-        @Nullable List<RoleDescriptor> roleDescriptors
+        @Nullable List<RoleDescriptor> roleDescriptors,
+        @Nullable List<RoleDescriptor> limitedByRoleDescriptors
+    ) {
+        this(
+            name,
+            id,
+            creation,
+            expiration,
+            invalidated,
+            username,
+            realm,
+            metadata,
+            roleDescriptors,
+            limitedByRoleDescriptors == null ? null : new RoleDescriptorsIntersection(List.of(limitedByRoleDescriptors))
+        );
+    }
+
+    private ApiKey(
+        String name,
+        String id,
+        Instant creation,
+        Instant expiration,
+        boolean invalidated,
+        String username,
+        String realm,
+        @Nullable Map<String, Object> metadata,
+        @Nullable List<RoleDescriptor> roleDescriptors,
+        @Nullable RoleDescriptorsIntersection limitedBy
```

Comment:
I added this new `RoleDescriptorsIntersection` mainly for future proof (in case we support derived keys properly). It also mirrors what we did with [RoleReferenceIntesection](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/RoleReferenceIntersection.java) and [LimitedRole](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/LimitedRole.java).
It does also introduce some code level overhead. Let me know if you dislike it and prefer a simpler `List<RoleDescriptor>`.

Reply:
Missed this comment. `RoleDescriptorsIntersection` makes sense to me. We know we will need something like this and all the work is already there now 👍